assembly/src/errors.rs

```diff
@@ -1,12 +1,32 @@
 use alloc::{string::String, sync::Arc, vec::Vec};
-use vm_core::mast::MastForestError;
+use vm_core::{errors::KernelError, mast::MastForestError};
 
 use crate::{
-    ast::{FullyQualifiedProcedureName, ProcedureName},
+    ast::FullyQualifiedProcedureName,
     diagnostics::{Diagnostic, RelatedError, RelatedLabel, Report, SourceFile},
-    KernelError, LibraryNamespace, LibraryPath, RpoDigest, SourceSpan,
+    LibraryNamespace, LibraryPath, RpoDigest, SourceSpan,
 };
 
+// ASSEMBLER ERROR
+// ================================================================================================
+
+/// An error generated during instantiation of an [super::Assembler].
+#[derive(Debug, thiserror::Error, Diagnostic)]
+#[non_exhaustive]
+pub enum AssemblerError {
```

Comment:
Same question. Taking it a step further, I would suggest that we move away from error types like these, and lean more heavily on `Report`, and the various related means of constructing diagnostics. Not only does this allow us to represent a far greater variety of errors, but we can be more precise when reporting errors, rather than needing to add a new variant to an enum, or try to contort an existing variant to meet our needs.

Thanks to `assert_diagnostic` and friends, we can also be more precise when asserting that specific failures are caught in our tests - rather than exhaustively match every detail of an error type, we can assert things about what an end user would see reported to them.

Reply:
I should note that our existing use of `AssemblyError` could go away in favor of just using `Report`, with maybe a handful of specific error types where useful internally. The only reason it still exists is basically because I was too lazy to update all the places where it was used when I did the assembler refactoring.

Reply:
I removed the `AssemblerError` as it is no longer needed after adding `KernelLibrary` module.